FAT-1889

automaticqa opened the pull request:
Создано:
	(1) БД пород и обслуживающий её класс DogKindsSQLiteDataSource. 
	(2) Класс DogKindSourceCoordinator имплеметирующий IBreedDataSource с методами getDogsKinds, getBreedsImage, updateBreedDBWithUriImage. В зависимости от усливий, он использует или методы WebBreedsDataSource, или DogKindsSQLiteDataSource.
	(3) Класс-синглтон BreedPicasso и метод, возвращающий экземпляр Picasso. 
	При переходе в DogsKindListActivity идет проверка: метод DogKindSourceCoordinator'a getDogsKinds, если БД пород пустая, то делаем запрос к API, вытаскиваем List<Пород> и создаем БД, при этом у каждой записи поле imageString="". Затем в адаптере пород проверяем, если поле пустое тогда делаем запрос к API и update'им данное поле в БД

automaticqa committed:
FAT_1858_Picasso
Добавлен getBreedsImage метод

automaticqa committed:
FAT_1858_Picasso
Вытащил uri c помощью метода getBreedImageUriString

automaticqa committed:
FAT_1858_Picasso
Создана DogKindTable

automaticqa committed:
FAT_1858_Picasso
Создан DogKindsSQLiteDataSource

automaticqa committed:
FAT_1889
Добавлен в модель DogKind id

automaticqa committed:
FAT_1889
В колбэк, который дергается в адаптере, добавлена реализация Picasso.

automaticqa committed:
FAT-1889
Создан интерфейс IResponseBreedCallback<String, ImageView, DogKind> и его реализация в DogsKindAdapter'e

automaticqa committed:
FAT-1889
временно добавлена активити TempActivity

automaticqa committed:
FAT-1889
метод addDogKind в DogKindsSQLiteDataSource работает корректно

automaticqa committed:
FAT-1889
добавил переменную dogBreeds в MyApp, там же сразу и инициализирую из API

automaticqa committed:
FAT-1889
DogKindSource теперь отвечает за List<DogKind> и имплементит IBreedDataSource

automaticqa committed:
FAT-1889
Добавлен метод isBreedDatabaseEmpty

automaticqa committed:
FAT-1889
Добавлен interface IDatabaseCallback.
Изменён метод getDogKinds

automaticqa committed:
FAT-1889
Добавлен удалённый ранее DogKindListGenerator

automaticqa committed:
FAT-1889
Исправил setData

automaticqa committed:
FAT-1889
Проведена замена класса WebBreedsDataSource на DogKindLocalDataSource

automaticqa committed:
FAT-1889
Добавлена картинка placeholder

automaticqa committed:
FAT-1889
Исправлен недочёт в работе метода loadDogKinds,в процесcе чтения из БД добавлена инициализация id-поля при создании списка пород.

automaticqa committed:
FAT-1889
Добавлена проверка в адаптере: если поле uriImageString экз-ра DogKind не пусто, то в БД лезть не нужно. Если пусто, то лезем в БД и обновляем данное поле у соответствующей породы.

automaticqa committed:
FAT-1889
Решена проблемa с установкой uri в imageView

automaticqa committed:
FAT-1889
DogsListActivity.
Переделана логика установки картинки в DogKindListAcitivty и DogListActivity. В зависимости от того фотография ли в поле uriImageString или drawable ресурс: переделан метод DogImageUtils.getDogImage()

automaticqa committed:
FAT-1889
1. Исправил ошибку в path в DogBreedsAPI.
"dogKindInstance" вернул обратно "breed".
2. Заменил DogsKindsLocalDataSource на WebBreedsDataSource

automaticqa committed:
FAT-1889
Добавлен ImageLoader класс

automaticqa committed:
FAT-1889
Добавлен Target переменная но не работает пока

automaticqa committed:
FAT-1889
Исправил ошибку. Записывал в uri url

automaticqa committed:
FAT-1889
Добавил targer в imageView

automaticqa committed:
FAT-1889
DogsKindsListActivity: Метод onResume. Убрал уже не нужную проверку

automaticqa committed:
FAT-1889
DogsKindAdapter. Добавлен ProgressBar

automaticqa committed:
FAT-1889
Добавил кастомный Picasso

automaticqa committed:
FAT-1889
BreedPicasso. Добавил методы initWithTarget и intoImageView

automaticqa committed:
FAT-1889
BreedPicasso. Добавил метод getTarget

automaticqa committed:
FAT-1889
ProgressBar отрегулирован

automaticqa committed:
FAT-1889
Удалил ненужный progressbar

automaticqa committed:
FAT-1889
Перенес инициализацию DBHelper'ом DatabaseManager'a из OwnersSQLiteDataSource в MyApp

automaticqa committed:
FAT-1889
Удалил еще комменты

automaticqa committed:
FAT-1889
Удалил еще комменты

automaticqa committed:
Merge branch 'master' into FAT-1889

# Conflicts:
#	app/src/main/java/com/andersen/dogsapp/dogs/MyApp.java
#	app/src/main/java/com/andersen/dogsapp/dogs/data/web/WebBreedsDataSource.java
#	app/src/main/java/com/andersen/dogsapp/dogs/ui/dogskinds/DogsKindsListActivity.java

automaticqa committed:
FAT-1889
DogsInfoActivity проверка содержимого поля uriImageString

automaticqa committed:
FAT-1889
Установил resize 500 width

automaticqa committed:
FAT-1889
Удалил лишние строки в getTarget

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/DogKindSourceCoordinator.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    public void getDogsKinds(IWebCallback<List<DogKind>> responseCallback, IDatabaseCallback<List<DogKind>> dbCallback) {
```
логика распределения с какого датасоурса брать инфу должен заниматься репозиторий, вообще у нас один единый репозиторий сейчас, а обычно для каждой сущности выделяют свой репозиторий, а уже общую работу с несколькими репозиториями - в интеракторы(бизнес-логика), а с интеракторами уже работает презентер и тд и тп, в нашем случае засунь это в DataRepository напрямую

automaticqa committed:
FAT-1889
Исправил try в SQLDataSource'ах

automaticqa committed:
FAT-1889
Вернул иниц-цию dbHelper обратно в OwnerSQL..Source

automaticqa committed:
FAT-1889
AndroidManifest allowBackup="false"

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/ui/dogskinds/DogsKindsListActivity.java:
```diff
@@ -56,23 +62,55 @@ protected void onSaveInstanceState(Bundle outState) {
-            DataRepository.get().getDogKinds(dogBreeds -> {
-                dogKinds = dogBreeds;
-                runOnUiThread(this::updateUI);
+        DataRepository.get().getDogKinds(dogBreeds -> {
```
)))) тут достаточно одного колбэка, и твоя задача в него передать List<DogKind>

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/web/imageloader/BreedPicasso.java:
```diff
@@ -0,0 +1,90 @@
+
+public class BreedPicasso {
+    private static BreedPicasso instance;
+    final int placeholder = R.drawable.smiled_dog_face;
```
почему это и ниже не private?

dmitrybubnovsky replied:
повторяю вопрос

automaticqa committed:
FAT-1889
Вернул иниц-цию dbHelper'a обрвтно в MyApp

automaticqa committed:
FAT-1889
Переименован *Coordinator в DogKindRepository

automaticqa committed:
FAT-1889
Добавил OwnersDataRepository

automaticqa committed:
FAT-1889
Добавлен DogsDataRepository

automaticqa committed:
FAT-1889
Добавлен паккет repositories

automaticqa committed:
FAT-1889
Переделан метод getDogKinds - вместо двух колбэков - один

automaticqa committed:
FAT-1889
OwnersDataRepository содержит поле сответств-го интерфейса

automaticqa committed:
FAT-1889
DogsDataRepository содержит поле сответств-го интерфейса

automaticqa committed:
FAT-1889
убрал использование DataRepository в работе с моделями Owner и Dogs

automaticqa committed:
FAT-1889
OwnersRepository: метод getOwners исправле

automaticqa committed:
FAT-1889
Dogs/OwnersRepository ничего не имплементят

automaticqa committed:
FAT-1889
DogsKindListActivity: onResume: замена DataRepository на BreedsRepository

automaticqa committed:
FAT-1889
BreedsRepository: убрал поле интерфейса

automaticqa committed:
FAT-1889
Ctrl+Alt+L

automaticqa committed:
FAT-1889
updateBreedDBWithUriImage возвращает void

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/repositories/DogsRepository.java:
```diff
@@ -0,0 +1,39 @@
+    private static DogsRepository instance;
+    private IDogsDataSource iDogsDataSource;
+
+    public DogsRepository(IDogsDataSource iDogsDataSource) {
```
там где синглтон, конструктор - приватный

StanlyT replied:
ну да в этом же и весь смысл, 
пропустил... и не только здесь, исправляю

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/repositories/DogsRepository.java:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    public static DogsRepository get() {
```
везде пройдись и перепиши на getInstance() а то не понятно

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/ui/dogs/DogsInfoActivity.java:
```diff
@@ -34,8 +35,13 @@ protected void onCreate(Bundle savedInstanceState) {
-        dogsPhoto.setImageDrawable(DogImageUtils.getDogImage(context, dog.getDogImageString()));
+        ImageView dogsImageView = findViewById(R.id.dog_imageview);
+        String dogImageString = dog.getDogImageString();
+        if (dogImageString.contains("_doggy_dogg.jpg")) {
```
а можешь эту проверку вынести в DogImageUtils.hasNetworkImage(Dog dog)

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/ui/dogskinds/DogsKindsListActivity.java:
```diff
@@ -56,23 +60,50 @@ protected void onSaveInstanceState(Bundle outState) {
+
+        if (dogKindInstance.getUriImageString().isEmpty()) {
+            BreedsRepository.getInstance().getBreedsImage(dogKindString, uriBreedString -> {
+                final File breedImageFile = getImageBreedFile(getApplicationContext(), dogKindString);
```
все что в лямбде вынеси в отдельный метод

StanlyT replied:
вынес в отдельный метод, всё кроме BreedsRepository.getInstance().updateBreedDBWithUriImage( ). А то тогда этот метод выполнял две функции и название у него было бы странное. А так получилось saveBreedImageSetToView()

dmitrybubnovsky commented on app/src/main/res/layout/temp_layout_form.xml:
```diff
@@ -0,0 +1,79 @@
+                android:layout_height="wrap_content"
+                android:layout_margin="@dimen/margin_8dp"
+                android:text="Обновить БД"/>
+
```
пустые строки лишние

StanlyT replied:
этот лэйаут вообще я забыл удалить, это от тестовой активити)

automaticqa committed:
FAT-1889
Заменил get на getInstance.
Заменил try/finally на try/with resources
Исправил конструкторы-синглтоны на private

automaticqa committed:
FAT-1889
Вынес в отдельный метод saveBreedImageSetToView()

automaticqa committed:
FAT-1889
Удален лишний xml.
 Переименованы dogkind в breeds

automaticqa committed:
FAT-1889
Переименованы dogkind в breed

automaticqa committed:
FAT-1889
Переименованы еще dogKinds в breeds

automaticqa committed:
FAT-1889
Ctrl+Alt+L before push

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/database/BreedsSQLiteDataSource.java:
```diff
@@ -0,0 +1,97 @@
+        dbCallback.onResult(breeds);
+    }
+
+    public boolean isBreedsDatabaseEmpty() {
```
и так понятно что ты вызываешь метод у BreedsRepositroy я думаю тут можно проще назвать метод isDbEmpty() например

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/database/BreedsSQLiteDataSource.java:
```diff
@@ -0,0 +1,97 @@
+        DatabaseManager.getInstance().closeDB();
+    }
+
+    private void insertBreedInLoop(Breed breed) {
```
в начале public потом private методы, следи за этим, потому что когда человек заходит читать код, вначале смотрит на публичный контракт, а если ему надо уже идет смотреть приват методы

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/web/imageloader/BreedPicasso.java:
```diff
@@ -0,0 +1,79 @@
+        picasso.get()
+                .load(uriBreedString)
+                .placeholder(placeholder)
+                .resize(500,0)
```
что такое 500?

dmitrybubnovsky commented on app/src/main/java/com/andersen/dogsapp/dogs/data/web/imageloader/BreedPicasso.java:
```diff
@@ -0,0 +1,79 @@
+
+            @Override
+            public void onPrepareLoad(Drawable placeHolderDrawable) {
+                if (placeHolderDrawable != null) {
```
в методах в которых нет реализации пиши комменты //no implementation

automaticqa committed:
FAT-1889
Переименовал в isDbEmpty(). Изменил порядок следования методов, создал константы для targetWidth/Height
